docs/guide/configuring-rules-via-php-attributes.md

```diff
@@ -0,0 +1,611 @@
+            ]),
+        ]),
+    ]),
+];
+```
+
+it can be applied to DTO classes like this achieving the same effect:
+
+```php
+use Yiisoft\Validator\Rule\Each;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Nested;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\Rule\Url;
+
+final class Post
+{
+    #[HasLength(min: 1, max: 255)]
+    public string $title;
+
+    // "Nested" can be used without arguments, but make sure to fill the value with the instance in this case (here it's
+    // filled right in the constructor).
+    #[Nested]
+    public Author|null $author = null;
+
+    // Passing instances is available only since PHP 8.1.
+    #[Each(new Nested(File::class))]
+    public array $files = [];
+
+    public function __construct()
+    {
+        $this->author = new Author();
+    }
+}
+
+final class Author
+{
+    #[Required]
+    #[HasLength(min: 1, max: 50)]
+    public string $name;
+
+    #[Number(integerOnly: true, min: 18, max: 100)]
+    public int $age;
+}
+
+// Some rules, like "Nested" can be also configured through the class attribute.
+
+#[Nested(['url' => new Url()])]
+final class File
+{
+    public string $url;
+}
+```
+
+For better understanding of relations concept, it's recommended to read [Nested] and [Each] guides.
+
+## Traits
+
+Attributes can be used in traits as well. It might come in handy for reusing the same set of properties with identical 
+rules: 
+
+```php
+use Yiisoft\Validator\Rule\HasLength;
+
+trait TitleTrait
+{
+    #[HasLength(max: 255)]
+    public string $title;
+}
+
+final class BlogPost
+{
+    use TitleTrait;
+}
+
+final class WikiArticle
+{
+    use TitleTrait;
+}
+```
+
+## Inheritance
+
+Inheritance is supported, but there are some things to keep in mind:
+
+```php
+use Yiisoft\Validator\Rule\BooleanValue;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+
+class Car
+{
+    #[Required]
+    #[HasLength(min: 1, max: 50)]
+    public string $name;
+    
+    #[Required]
+    #[BooleanValue]
+    public $used;
+    
+    #[Required]
+    #[Number(max: 2000)]
+    public float $weight;     
+}
+
+class Truck extends Car
+{       
+    public string $name;
+    
+    #[Number(max: 3500)]
+    public float $weight;      
+}
+```
+
+In this case the set of rules for `Truck` will be:
+
+```php
+use Yiisoft\Validator\Rule\BooleanValue;
+use Yiisoft\Validator\Rule\Required;
+
+[
+    'used' => [
+        new Required(),
+        new BooleanValue(),
+    ],
+    'weight' => [
+        new Number(max: 3500),
+    ],
+];
+```
+
+So, to sum up:
+
+- Parent rules for overridden properties are ignored completely, only the ones from the child class are obtained.
+- All parent rules for properties not overridden in the child class are obtained fully.
+
+As for the data, default values set in the child class take precedence.
+
+## Adding attributes support to custom rules
```

Comment:
Maybe leave it here because it covers both `Composite` custom rules and regular custom rules (otherwise it needs to be split to 2 different files). And it's related with attributes.

Reply:
I think leaving it as is good.

Reply:
Discussed in the group chat with @samdark, it's OK.